R/geom_params.R

```diff
@@ -0,0 +1,187 @@
+#'   Named values or [character] strings.
+#'   Unnamed arguments will check whether any value was set for that parameter.
+#'   Named arguments will check whether the parameter with the same name has a
+#'   matching value.
+#'   Each argument should have a name matching a [ggplot][ggplot2::ggplot]
+#'   layer parameter.
+#'   Values may be passed as arguments or as list elements.
+#' @param params A named list of geom or stat parameter values, e.g.
+#'   \code{list(outlier.alpha = 0.01)}.
+#'   This list is combined with any inputs to `...`
+#' @param i A numerical index, e.g. \code{1}.
+#'
+#' @return A named logical vector of the same length as the number of inputs
+#'   to `...`.
+#' @family functions for checking geom parameters
+#' @export
+#'
+#' @examples
+#' require(ggplot2)
+#'
+#' p <- ggplot(data = diamonds, aes(x = cut, y = price)) +
+#'   geom_boxplot(varwidth = TRUE, outlier.alpha = 0.01, fill = "blue")
+#'
+#' uses_geom_params(
+#'   p, "boxplot", list(varwidth = TRUE, outlier.alpha = 0.01, fill = "blue")
+#' )
+#'
+#' uses_geom_params(
+#'   p, "boxplot", varwidth = TRUE, outlier.alpha = 0.01, fill = "blue"
+#' )
+#'
+#' # Unnamed arguments check that a parameter is set to any value
+#' uses_geom_params(p, "boxplot", "fill")
+uses_geom_params <- function(p, geom, ..., params = NULL, i = NULL) {
+  stop_if_not_ggplot(p)
+
+  layer <- get_geom_layer(p, geom = geom, i = i)$layer
+
+  params <- c(params, capture_dots(...))
+  named  <- names(params) != ""
+
+  user_params         <- names(params)
+  user_params[!named] <- as.character(params[!named])
+
+  default_params <- purrr::map_lgl(params, inherits, ".default_params")
+  params[default_params] <- purrr::map(
+    names(params)[default_params],
+    ~ unlist(unname(default_params(p, geom, ., i = i)))
+  )
+
+  result        <- logical(length(params))
+  names(result) <- user_params
+
+  user_params[user_params == "color"] <- "colour"
+
+  # Collect geom, stat, and aes parameters
+  all_params <- c(layer$geom_params, layer$stat_params, layer$aes_params)
+
+  # Add inherited default parameters
+  default_params <- default_params(p, geom)
+  inherited <- !names(default_params) %in% names(all_params)
+  all_params_with_inherited <- c(all_params, default_params[inherited])
+
+  result[named] <- purrr::map2_lgl(
+    params[named], all_params_with_inherited[user_params][named], identical
+  )
+  result[!named] <- user_params[!named] %in% names(all_params)
+  result
+}
+
+#' @rdname uses_geom_params
+#' @export
+uses_geom_param <- uses_geom_params
+
+#' What are the default parameters for a plot layer?
+#'
+#' @examples
+#' require(ggplot2)
+#'
+#' p <- ggplot(data = mpg, mapping = aes(x = displ, y = hwy)) +
+#'   geom_smooth(aes(color = class))
+#'
+#' # Returns the parameters the ggplot would use by default for a layer
+#' default_params(p, "smooth", "linetype")
+#' default_params(p, "smooth", "se", "level")
+#' default_params(p, "smooth")
```

Comment:
`default_params()` vs `default_label()` doesn't feel consistent and it'd be good to articulate why the difference exists or what benefit we derive out of the inconsistent naming that makes it worth it.

I also like `capture_dots()` but I'm starting to feel like it's bordering on an anti-pattern. I think I'd prefer

```
default_params(p, "smooth", c("se", "level"))
```

instead, where there are clearly three arguments, each of which can be named.

Reply:
Resolved in 670a993 (`params` instead of `...`) and f2b5d3f (renaming functions)